pkg/controller/sync/version/manager.go

```diff
@@ -276,146 +263,140 @@ func (m *VersionManager) versionQualifiedName(qualifiedName util.QualifiedName)
-	// caller (this function) whose actions will result in metadata
-	// changing.
-	metaAccessor, err := meta.Accessor(obj)
+	resourceVersion, err := getResourceVersion(obj)
 	if err != nil {
-		runtime.HandleError(errors.Wrapf(err, "Failed to retrieve meta accessor for %s %q", adapterType, key))
-		return util.StatusError
+		return errors.Wrapf(err, "Failed to retrieve resource version for %s %q", adapterType, key)
 	}
-
-	// Locking shouldn't be required here since there is only a single
-	// caller (this function) whose actions will result in metadata
-	// changing.
-	creationNeeded := len(metaAccessor.GetResourceVersion()) == 0
-
-	if creationNeeded {
-		// Make a copy to avoid mutating the map outside of a lock.
-		createdObj := obj.DeepCopyObject()
-		err := m.client.Create(context.TODO(), createdObj)
-		if apierrors.IsAlreadyExists(err) {
-			// Version was written to the API after the version manager loaded
-			glog.V(4).Infof("Refreshing %s %q from the API due to already existing", adapterType, key)
-			err := m.refreshVersion(obj)
+	refreshVersion := false
+	// TODO(marun) Centralize polling interval and duration
+	waitDuration := 30 * time.Second
+	err = wait.PollImmediate(100*time.Millisecond, waitDuration, func() (bool, error) {
+		if refreshVersion {
+			// Version was written to the API by another process after the last manager write.
+			var err error
+			resourceVersion, err = m.getResourceVersionFromAPI(qualifiedName)
 			if err != nil {
-				runtime.HandleError(errors.Wrapf(err, "Failed to refresh existing %s %q from the API", adapterType, key))
-				return util.StatusError
+				runtime.HandleError(errors.Wrapf(err, "Failed to refresh resource version %s %q from the API", adapterType, key))
+				return false, nil
 			}
-			return util.StatusNeedsRecheck
+			refreshVersion = false
 		}
-		if err != nil {
-			runtime.HandleError(errors.Wrapf(err, "Failed to create version %s %q", adapterType, key))
-			return util.StatusError
+
+		if resourceVersion == "" {
+			// Version resource needs to be created
+
+			createdObj := obj.DeepCopyObject()
+			err := setResourceVersion(createdObj, "")
+			if err != nil {
+				runtime.HandleError(errors.Wrapf(err, "Failed to clear the resource version for %s %q", adapterType, key))
+				return false, nil
+			}
+
+			glog.V(4).Infof("Creating %s %q", adapterType, qualifiedName)
+			err = m.client.Create(context.TODO(), createdObj)
+			if apierrors.IsAlreadyExists(err) {
+				glog.V(4).Infof("%s %q was created by another process.  Will refresh the resource version and attempt to update.", adapterType, qualifiedName)
+				refreshVersion = true
+				return false, nil
+			}
+			if err != nil {
+				runtime.HandleError(errors.Wrapf(err, "Failed to create version %s %q", adapterType, key))
+				return false, nil
+			}
+
+			// Update the resource version that will be used for update.
+			resourceVersion, err = getResourceVersion(createdObj)
```

Comment:
For the created resource, it seems no need to update again.
Do you trying to reuse the logic at the end of block to update local cached obj?

Reply:
The version detail is stored in the status of the resource.  Status is not set on creation - it needs to be set explicitly through the status subresource after creation.